klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt

```diff
@@ -60,7 +60,7 @@ data class PatternDateFormat @JvmOverloads constructor(
             }
             val chunk = s.readChunk {
                 val c = s.readChar()
-                while (s.hasMore && s.tryRead(c)) Unit
+                while (s.hasMore && (s.tryRead(c) || s.tryReadDaySuffix(c))) Unit
```

Comment:
Why is this change necessary? Is not covered multi-character patterns already somehow?

Reply:
`fun tryRead(expected: Char): Boolean {
		if (eof || peekChar() != expected) return false  
		readChar()  
		return true  
}`
In s.tryRead(current) if "expected" is not equal to the next character returns false and doesn't call to readChar(). 

`fun tryReadDaySuffix(current: Char): Boolean {
        if (current == 'd' && peekChar() == 'o') {
            readChar()
            return true
        }
        return false
}`

For this case I used tryReadDaySuffix(current) where if "current" is equal to "d" and next char is equal to "o" calls readChar()

Reply:
What about this?

```diff
Index: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt	(revision b730afdf20fa13b7792c4a078f5e3e81d78ce862)
+++ klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt	(date 1604673400838)
@@ -58,9 +58,9 @@
                     continue
                 }
             }
-            val chunk = s.readChunk {
+            val chunk = s.tryReadOrNull("do") ?: s.readChunk {
                 val c = s.readChar()
-                while (s.hasMore && (s.tryRead(c) || s.tryReadDaySuffix(c))) Unit
+                while (s.hasMore && (s.tryRead(c))) Unit
             }
             chunks.add(chunk)
         }
Index: klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt	(revision b730afdf20fa13b7792c4a078f5e3e81d78ce862)
+++ klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt	(date 1604673389034)
@@ -21,19 +21,15 @@
 		readChar()
 		return true
 	}
-    fun tryReadDaySuffix(current: Char): Boolean {
-        if (current == 'd' && peekChar() == 'o') {
-            readChar()
-            return true
-        }
-        return false
-    }
     fun tryRead(expected: String): Boolean {
         if (expected.length > available) return false
         for (n in expected.indices) if (this.str[offset + n] != expected[n]) return false
         offset += expected.length
         return true
     }
+    fun tryReadOrNull(expected: String): String? {
+        return if (tryRead(expected)) expected else null
+    }
     fun read(count: Int): String = this.str.substring(offset, (offset + count).coerceAtMost(length)).also { this.offset += it.length }
 	fun readRemaining(): String = read(available)
     fun readInt(count: Int): Int = read(count).toInt()
```

I think that tryReadDaySuffix doesn't belong to MicroStrReader. That class is not coupled at all to klock, it is a generic StrReader.
That function could be provided as a private extension function, but I think you don't want to try to read the day suffix on each iteration.

Reply:
In fact I think that would read better with this:

```
private fun MicroStrReader.readRepeatedChar(): String {
    return readChunk {
        val c = readChar()
        while (hasMore && (tryRead(c))) Unit
    }
}
```

```
chunks.add(s.tryReadOrNull("do") ?: s.readRepeatedChar())
```

Because I had to figure out what that loop did when I read it since the time it was written

Reply:
Done